io/zenoh-links/zenoh-link-shm/src/unix/unicast.rs

```diff
@@ -0,0 +1,573 @@
+    let _uplink = PipeR::new(&path_uplink, access_mode).await?; // uplink would be dropped, that is OK!
+
+    Ok((downlink, suffix, path_downlink, path_uplink))
+}
+
+async fn dedicate_pipe(
+    path_uplink: &str,
+    path_downlink: &str,
+    access_mode: u32,
+) -> ZResult<(PipeR, u32, String, String)> {
+    for _ in 0..LINUX_PIPE_DEDICATE_TRIES {
+        match create_pipe(path_uplink, path_downlink, access_mode).await {
+            Err(_) => {}
+            val => {
+                return val;
+            }
+        }
+    }
+    bail!("Unabe to dedicate pipe!")
+}
+
+struct UnicastPipeClient;
+impl UnicastPipeClient {
+    async fn connect_to(endpoint: EndPoint) -> ZResult<UnicastPipe> {
+        let (path_uplink, path_downlink, access_mode) = parse_pipe_endpoint(&endpoint);
+
+        // open the request channel
+        // this channel would be used to invite listener to the dedicated channel
+        // listener owns the request channel, so failure of this call means that there is nobody listening on the provided endpoint
+        let mut request_channel = PipeW::new(&path_uplink).await?;
+
+        // create dedicated channel prerequisities. The creation code also ensures that nobody else would use the same channel concurrently
+        let (
+            mut dedicated_downlink,
+            dedicated_suffix,
+            dedicated_donlink_path,
+            dedicated_uplink_path,
+        ) = dedicate_pipe(&path_uplink, &path_downlink, access_mode).await?;
+
+        // invite the listener to our dedicated channel over the requet channel
+        Invitation::send(dedicated_suffix, &mut request_channel).await?;
+
+        // read responce that should be sent over the dedicated channel, confirming that everything is OK
+        // on the listener's side and it is already working with the dedicated channel
+        Invitation::expect(dedicated_suffix, &mut dedicated_downlink).await?;
+
+        // open dedicated uplink
+        let mut dedicated_uplink = PipeW::new(&dedicated_uplink_path).await?;
+
+        // final confirmation over the dedicated uplink
+        Invitation::confirm(dedicated_suffix, &mut dedicated_uplink).await?;
+
+        // create Locators
+        let local = Locator::new(
+            endpoint.protocol(),
+            dedicated_donlink_path,
+            endpoint.metadata(),
+        )?;
+        let remote = Locator::new(
+            endpoint.protocol(),
+            dedicated_uplink_path,
+            endpoint.metadata(),
+        )?;
+
+        Ok(UnicastPipe {
+            r: UnsafeCell::new(dedicated_downlink),
+            w: UnsafeCell::new(dedicated_uplink),
+            local,
+            remote,
+        })
+    }
+}
+
+struct UnicastPipe {
+    r: UnsafeCell<PipeR>,
+    w: UnsafeCell<PipeW>,
+    local: Locator,
+    remote: Locator,
+}
+
+impl UnicastPipe {
+    #[allow(clippy::mut_from_ref)]
+    fn get_r_mut(&self) -> &mut PipeR {
```

Comment:
Add `SAFETY` comment on what are the assumptions that are held in order for the `unsafe` block to actually be safe.